https://adventofcode.com/2022/day/10

X value set by input
noop = nothing
addx +/-num (takes two cycles to affect)


In [38]:
input = '''noop
addx 3
addx -5
noop
noop'''

In [18]:
for line in input.split('\n'):
    print(line)
    if 'noop' in line:
        val = 0
    else:
        _, val = line.split( )
    print(val)

noop
0
addx 3
3
addx -5
-5
noop
0
noop
0


In [5]:
line = 'noop'
'noop' in line

True

In [7]:
line = 'addx 3'
_, val = line.split( )
val

'3'

In [42]:
now, next1,next2  = 0,0,0
x = 1
cycle = 1
for line in input.split('\n'):
    #print(line)
    if 'noop' in line:
        next2 = 0
        x += now
        print('noop')
    else:
        _, val = line.split( )
        next2 = int(val)
        x += now
        print(f"addx {val}")
    print(f"cycle: {cycle}, now:{now}, next:{next1}, next2:{next2}, end cycle xval: {x}")
    now = next1
    next1 = next2
    cycle += 1

noop
cycle: 1, now:0, next:0, next2:0, end cycle xval: 1
addx 3
cycle: 2, now:0, next:0, next2:3, end cycle xval: 1
addx -5
cycle: 3, now:0, next:3, next2:-5, end cycle xval: 1
noop
cycle: 4, now:3, next:-5, next2:0, end cycle xval: 4
noop
cycle: 5, now:-5, next:0, next2:0, end cycle xval: -1


This is close to the desc:

* At the start of the first cycle, the noop instruction begins execution. During the first cycle, X is 1. After the first cycle, the noop instruction finishes execution, doing nothing. `So X = 1`
* At the start of the second cycle, the addx 3 instruction begins execution. During the second cycle, X is still 1.
* During the third cycle, X is still 1. After the third cycle, the addx 3 instruction finishes execution, setting X to 4. `my code has x=1`
* At the start of the fourth cycle, the addx -5 instruction begins execution. During the fourth cycle, X is still 4.
* During the fifth cycle, X is still 4. After the fifth cycle, the addx -5 instruction finishes execution, setting X to -1.

I don't understand the description above -- so let's try the longer input to see what happens. . .

In [76]:
input = '''addx 15
addx -11
addx 6
addx -3
addx 5
addx -1
addx -8
addx 13
addx 4
noop
addx -1
addx 5
addx -1
addx 5
addx -1
addx 5
addx -1
addx 5
addx -1
addx -35
addx 1
addx 24
addx -19
addx 1
addx 16
addx -11
noop
noop
addx 21
addx -15
noop
noop
addx -3
addx 9
addx 1
addx -3
addx 8
addx 1
addx 5
noop
noop
noop
noop
noop
addx -36
noop
addx 1
addx 7
noop
noop
noop
addx 2
addx 6
noop
noop
noop
noop
noop
addx 1
noop
noop
addx 7
addx 1
noop
addx -13
addx 13
addx 7
noop
addx 1
addx -33
noop
noop
noop
addx 2
noop
noop
noop
addx 8
noop
addx -1
addx 2
addx 1
noop
addx 17
addx -9
addx 1
addx 1
addx -3
addx 11
noop
noop
addx 1
noop
addx 1
noop
noop
addx -13
addx -19
addx 1
addx 3
addx 26
addx -30
addx 12
addx -1
addx 3
addx 1
noop
noop
noop
addx -9
addx 18
addx 1
addx 2
noop
noop
addx 9
noop
noop
noop
addx -1
addx 2
addx -37
addx 1
addx 3
noop
addx 15
addx -21
addx 22
addx -6
addx 1
noop
addx 2
addx 1
noop
addx -10
noop
noop
addx 20
addx 1
addx 2
addx 2
addx -6
addx -11
noop
noop
noop'''

In [ ]:
now, next1, next2  = 0,0,0
x = 1
cycle = 1
for line in input.split('\n'):
    #print(line)
    if 'noop' in line:
        next2 = 0
        x += now
        #print('noop')
    else:
        _, val = line.split( )
        next2 = int(val)
        x += now
        #print(f"addx {val}")
    print(f"cycle: {cycle}, now:{now}, next:{next1}, next2:{next2}, xval: {x}")
    now = next1
    next1 = next2
    cycle += 1
    if cycle % 20 == 0:
        print(f"cycle: {cycle}, now:{now}, next:{next1}, next2:{next2}, xval: {x}")

So this suggests a different way to solve the problem:
* During the 20th cycle, register X has the value 21, so the signal strength is 20 * 21 = 420. (The 20th cycle occurs in the middle of the second addx -1, so the value of register X is the starting value, 1, plus all of the other addx values up to that point: 1 + 15 - 11 + 6 - 3 + 5 - 1 - 8 + 13 + 4 = 21.)

count cycles (addx += 2, noop += 1) to keep track of the cycles
sum the values

In [43]:
cycle = 1
x = 1
for line in input.split('\n'):
    #print(line)
    if 'noop' in line:
        cycle += 1
    else:
        _, val = line.split( )
        x += int(val)
        cycle += 2
    print(f"cycle:{cycle} x value:{x}")


cycle:2 x value:1
cycle:4 x value:4
cycle:6 x value:-1
cycle:7 x value:-1
cycle:8 x value:-1


try this on the longer input. . . 

In [53]:
cycle = 1
x = 1
cycle_list = [19,20,59,60,99,100,139,140,179,180,219,220]
for line in input.split('\n'):
    #print(line)
    if 'noop' in line:
        cycle += 1
    else:
        _, val = line.split( )
        x += int(val)
        cycle += 2
    if cycle in cycle_list:
        print(f"cycle:{cycle}, xval:{x}")



cycle:19, xval:21
cycle:20, xval:21
cycle:59, xval:19
cycle:60, xval:19
cycle:100, xval:18
cycle:140, xval:21
cycle:179, xval:16
cycle:180, xval:16
cycle:219, xval:18


In [65]:
cycle = 1
x = 1
cycle_list = [19,20,59,60,99,100,139,140,179,180,219,220]
val_list = []
for line in input.split('\n'):
    #print(cycle)
    if 'noop' in line:
        cycle += 1
    else:
        _, val = line.split( )
        x += int(val)
        cycle += 2
    if cycle in cycle_list:
        print(f"cycle:{cycle}, xval:{x}")
        val_list.append(x)
multipliers = [20,60,100,140,180,220]
sum = 0
for i in range(0,len(multipliers)):
    sum += val_list[i] * multipliers[i]
    print(f"multiplier:{multipliers[i]} value: {val_list[i]}")
    print(sum)

cycle:19, xval:21
cycle:20, xval:21
cycle:59, xval:19
cycle:60, xval:19
cycle:100, xval:18
cycle:140, xval:21
cycle:179, xval:16
cycle:180, xval:16
cycle:219, xval:18
multiplier:20 value: 21
420
multiplier:60 value: 21
1680
multiplier:100 value: 19
3580
multiplier:140 value: 19
6240
multiplier:180 value: 18
9480
multiplier:220 value: 21
14100


In [66]:
20*21 + 60*19 + 100 * 18 + 140 * 21 + 180 * 16 + 220 * 18

13140

In [67]:
wdir = '/home/alp/Google Drive/Python/AdventCode/2022'
with open (f"{wdir}/10input.txt") as f_in:
    input = f_in.read().split('\n')

In [68]:
cycle = 1
x = 1
cycle_list = [19,20,59,60,99,100,139,140,179,180,219,220]
for line in input:
    #print(cycle)
    if 'noop' in line:
        cycle += 1
    else:
        _, val = line.split( )
        x += int(val)
        cycle += 2
    if cycle in cycle_list:
        print(f"cycle:{cycle}, xval:{x}")


cycle:20, xval:21
cycle:59, xval:17
cycle:100, xval:21
cycle:139, xval:21
cycle:140, xval:21
cycle:179, xval:8
cycle:220, xval:21


In [69]:
cycle = 1
x = 1
cycle_list = [19,20,59,60,99,100,139,140,179,180,219,220]
val_list = []
for line in input:
    #print(cycle)
    if 'noop' in line:
        cycle += 1
    else:
        _, val = line.split( )
        x += int(val)
        cycle += 2
    if cycle in cycle_list:
        print(f"cycle:{cycle}, xval:{x}")
        val_list.append(x)
multipliers = [20,60,100,140,180,220]
sum = 0
for i in range(0,len(multipliers)):
    sum += val_list[i] * multipliers[i]
    print(f"multiplier:{multipliers[i]} value: {val_list[i]}")
    print(sum)


cycle:20, xval:21
cycle:59, xval:17
cycle:100, xval:21
cycle:139, xval:21
cycle:140, xval:21
cycle:179, xval:8
cycle:220, xval:21
multiplier:20 value: 21
420
multiplier:60 value: 17
1440
multiplier:100 value: 21
3540
multiplier:140 value: 21
6480
multiplier:180 value: 21
10260
multiplier:220 value: 8
12020


submitted 12020 -- but this is too low

let's do it manually and see if we come up with the same value

In [70]:
20 * 21 + 60 * 17 + 100 * 21 + 140 * 21 + 180 * 8 + 220 * 21

12540

OK, that's correct. So what's wrong with my code?

In [71]:
val_list

[21, 17, 21, 21, 21, 8, 21]

So the problem is that I'm double copying the 139/140 values.

Here's a clunky solution:

In [74]:
cycle = 1
x = 1
cycle_list = [(19,20),(59,60),(99,100),(139,140),(179,180),(219,220)]
val_list = []
for line in input:
    #print(cycle)
    if 'noop' in line:
        cycle += 1
    else:
        _, val = line.split( )
        x += int(val)
        cycle += 2
    for pair in cycle_list:
        if cycle in pair:
            print(f"cycle:{cycle}, xval:{x}")
            val_list.append(x)
multipliers = [20,60,100,140,180,220]
sum = 0
for i in range(0,len(multipliers)):
    sum += val_list[i] * multipliers[i]
    print(f"multiplier:{multipliers[i]} value: {val_list[i]}")
    print(sum)


cycle:20, xval:21
cycle:59, xval:17
cycle:100, xval:21
cycle:139, xval:21
cycle:140, xval:21
cycle:179, xval:8
cycle:220, xval:21
multiplier:20 value: 21
420
multiplier:60 value: 17
1440
multiplier:100 value: 21
3540
multiplier:140 value: 21
6480
multiplier:180 value: 21
10260
multiplier:220 value: 8
12020


That doesn't work. Problem is that on cycles 139 and 140 we have a "noop" so, the value for 139 and 140 are getting double entered.

So let's delete the pairs as they are found:

In [75]:
cycle = 1
x = 1
cycle_list = [(19,20),(59,60),(99,100),(139,140),(179,180),(219,220)]
val_list = []
for line in input:
    #print(cycle)
    if 'noop' in line:
        cycle += 1
    else:
        _, val = line.split( )
        x += int(val)
        cycle += 2
    for pair in cycle_list:
        if cycle in pair:
            print(f"cycle:{cycle}, xval:{x}")
            val_list.append(x)
            cycle_list = cycle_list[1:]
multipliers = [20,60,100,140,180,220]
sum = 0
for i in range(0,len(multipliers)):
    sum += val_list[i] * multipliers[i]
    print(f"multiplier:{multipliers[i]} value: {val_list[i]}")
    print(sum)


cycle:20, xval:21
cycle:59, xval:17
cycle:100, xval:21
cycle:139, xval:21
cycle:179, xval:8
cycle:220, xval:21
multiplier:20 value: 21
420
multiplier:60 value: 17
1440
multiplier:100 value: 21
3540
multiplier:140 value: 21
6480
multiplier:180 value: 8
7920
multiplier:220 value: 21
12540


Lets look at https://aoc.just2good.co.uk/2022/10 to see a more elegent solution.

Again, he uses classes (he loves!!!) classes

Part 2: x = midpoint of 3-pixel sprite (= x +/- 1)
input / 40 = # lines on the screen
point on row = range (0,40)
each cycle, pixel += 1 on row
if pixel in sprite, --> # else .

In [79]:
cycle = 1
pixel = 0
x = 1
sprite = [0,1,2]
for line in input.split('\n'):
    if 'noop' in line:
        pause = 0
        val = 0
    else:
        _, val = line.split( )
        pause = 1
        val = int(val)
    
    

24.475

In [91]:
def check_overlap(pixel,x):
    sprite = range(x-1,x+1)
    if pixel in sprite:
        return True
    else:
        return False

instructions = input.split('\n')
now, next = 0,0
x = 1
cycle = 1
pixel = 0
ins_num = 0
screen = []
screen_line = []
while pixel < 40:
#for pixel in range(0,40):
    inst = instructions[ins_num]
    if 'noop'in inst:
        if check_overlap(pixel,x):
            screen_line.append('#')
        else:
            screen_line.append('.')
        print(pixel,x)
        next = 0
        pixel += 1

    else:
        _, val = inst.split( )
        next = int(val)
        for i in range (0,2):
            x += now
            if check_overlap(pixel,x):
                screen_line.append('#')
            else:
                screen_line.append('.')
            print(pixel,x)
            now = next
            next = 0
            pixel += 1
    ins_num += 1

print(screen_line)




        


0 1
1 16
2 16
3 5
4 5
5 11
6 11
7 8
8 8
9 13
10 13
11 12
12 12
13 4
14 4
15 17
16 17
17 21
18 21
19 21
20 20
21 20
22 25
23 25
24 24
25 24
26 29
27 29
28 28
29 28
30 33
31 33
32 32
33 32
34 37
35 37
36 36
37 36
38 1
39 1
40 2
['#', '.', '.', '.', '#', '.', '.', '#', '#', '.', '.', '#', '#', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '.']


In [85]:
def check_overlap(pixel,x):
    #sprite = []
    sprite = range(x-1,x+1)
    if pixel in sprite:
        return True
    else:
        return False

check_overlap(0,12)

False

OK, this is a fucking mess. Let's see how Dazbo does it with classes

https://aoc.just2good.co.uk/2022/10

In [93]:
from pathlib import Path
import time

"""SCRIPT_DIR = Path(__file__).parent
# INPUT_FILE = Path(SCRIPT_DIR, "input/sample_input.txt")
INPUT_FILE = Path(SCRIPT_DIR, "input/input.txt")
"""

class CrtComputer():
    """ Performs instructions. Add instructions take 2 ticks. Noop takes 1 tick. """

    def __init__(self, instructions: list[str]) -> None:
        # self (required), instructions (input) as list of strings; 
        # --> is a function annotation that says fn will not return a value (None). This is not forced.
        self._x = 1 # represents middle of horizontal sprite position
        self._instructions = self._convert_to_instructions(instructions)
        self._ip = 0    # instruction pointer
        self.cycle = 0
        self._doing = []  # [[instruction], duration]
        self.running_program = True # Set to false when instructions are complete
    
    # variables with underscores self._x : underscore is HINT to other prorammer that varible is for internal use
    # https://dbader.org/blog/meaning-of-underscores-in-python
    # python does not enforce the rule about variables
    # it DOES treat module names with underscores differently: from my_module import * won't bring in functions that start with underscore

    @property               # decorator keyword
    def x(self):            # header - getter fn is names like property to be defined (x)
        return self._x      # getter function: get value and returns it 
    
    @property
    def signal_strength(self) -> int:
        return self.cycle * self.x
    
    # https://www.freecodecamp.org/news/python-property-decorator/
    # @property allows for three methods: getter, setter, deleter
    # 
    # these are both getter methos, since they return values
    # the underscore (_x) indicates the variable is internal and shouldn't be accessed outside the class
    # @x.setter     would be used as the name for the decorator setter function 
    # @x.deleter    would be used as the name for the decorator deleter function

    # underscore b/f function name -> internal to class
    # self, data are input. Based on the pror defn, I think the :list[str] says this what the data consists of
    # or is it saying what the input should be formatted like?
    def _convert_to_instructions(self, data: list[str]): 
        """ Create a list of instructions in [[instr, val],[],...] format. """
        instructions = []
        for line in data:
            line_words = line.split()
            instr = line_words[0]
            val = None
            if len(line_words) > 1:
                val = int(line_words[1])
            instructions.append((instr, val))
        return instructions

    def tick(self):
        """ Perform one CPU cycle """
        # print(self)
        # initial defn: self._doing = []  # [[instruction], duration]

        if len(self._doing) > 0: # we're processing an instruction
            self._doing[1] -= 1
            
            if self._doing[1] == 0:
                # Complete the running instruction
                instruction = self._doing[0]
                self.__getattribute__(f"_op_{instruction[0]}")(instruction)
                # this calls the OP fns (below) using 
                # print(f"Completed instruction: {instruction}")
                
                self._start_next_instruction() # and start the next one
        else: # our first instruction
            self._start_next_instruction()
            
        self.cycle += 1

    def _start_next_instruction(self):
        """ Takes an instruction, and calls the appropriate implementation method. """
        #         self._ip = 0    # instruction pointer (that is, where are we in the list of instructions?)
        instruction = self._instructions[self._ip]
        # print(f"Starting instruction: {instruction}")
        
        if instruction[0] == "addx":
            self._doing = [instruction, 2]
        elif instruction[0] == "noop":
            self._doing = [instruction, 1]

        self._ip += 1 # increment the instruction pointer
        if self._ip == len(self._instructions): # we've finished
            self.running_program = False

    def _op_addx(self, instruction: tuple):
        """ Takes 2 cycles. Adds val to register x """
        self._x += instruction[-1]

    def _op_noop(self, _: tuple):
        """ Takes 1 cycle. Does nothing. Instruction parameter will be empty. """
    
    def __repr__(self):
        return f"{self.__class__.__name__}(Cycle={self.cycle};x={self._x},pixel={self._display_posn})"

wdir = '/home/alp/Google Drive/Python/AdventCode/2022'
with open (f"{wdir}/10input.txt") as f_in:
    data = f_in.read().splitlines()

interesting_cycles = [20, 60, 100, 140, 180, 220]
signal_strength_sum = 0
crt_computer = CrtComputer(data)




In [96]:
while crt_computer.running_program:
    crt_computer.tick()
    if crt_computer.cycle in interesting_cycles:
        signal_strength_sum += crt_computer.signal_strength

print(signal_strength_sum)


12540


For part two, he just needs to 
1. add some a couple more properties to the class __init__ statement and a couple of constants (ALL CAP)
1. Chagne the tick method to add a screen update call
1. write an update display fn 

In [113]:
class CrtComputer():
    """ Performs instructions. Add instructions take 2 ticks. Noop takes 1 tick. """
    DISPLAY_WIDTH = 40
    DISPLAY_HEIGHT = 6
    LIT = "#"
    
    def __init__(self, instructions: list[str]) -> None:
        self._x = 1 # represents middle of horizontal sprite position
        self._instructions = self._convert_to_instructions(instructions)
        self._ip = 0    # instruction pointer
        self.cycle = 0
        self._doing = []  # [[instruction], duration]
        self.running_program = True # Set to false when instructions are complete
        
        self._display_posn = [0,0] # NEW: col, row for display
        self._display = [["." for _ in range(CrtComputer.DISPLAY_WIDTH)] 
                            for _ in range(CrtComputer.DISPLAY_HEIGHT)]
        # the two lines above are the initial settings for the pixel position and
        # the screen (set to blank = . for the width and height of the screen)
        # creates a list of six lists [[1][2][3][4][5][6]]

    @property
    def x(self):
        return self._x
    
    @property
    def signal_strength(self) -> int:
        return self.cycle * self.x
    
    def _convert_to_instructions(self, data: list[str]):
        """ Create a list of instructions in [[instr, val],[],...] format. """
        instructions = []
        for line in data:
            line_words = line.split()
            instr = line_words[0]
            val = None
            if len(line_words) > 1:
                val = int(line_words[1])
            instructions.append((instr, val))
                
        return instructions

    def tick(self):
        """ Perform one CPU cycle """
        # print(self)

        if len(self._doing) > 0: # we're processing an instruction
            self._doing[1] -= 1
            
            if self._doing[1] == 0:
                # Complete the running instruction
                instruction = self._doing[0]
                self.__getattribute__(f"_op_{instruction[0]}")(instruction)
                # print(f"Completed instruction: {instruction}")
                
                self._start_next_instruction() # and start the next one
        else: # our first instruction
            self._start_next_instruction()
        
        self._update_display() # NEW --> call the display update fn
            
        self.cycle += 1

    def _update_display(self):
        # Current horizontal pixel position being drawn
        x_posn = self._display_posn[0]
        
        # Check if horizontal position is *within* to current sprite position
        # The sprite is 3 pixels wide, and the x register gives us the middle position
        if x_posn in range(self.x-1, self.x+2):
            self._display[self._display_posn[1]][x_posn] = CrtComputer.LIT

        # display position moves across the row, then down, 
        # one pixel at a time with each tick        
        if x_posn < CrtComputer.DISPLAY_WIDTH-1:
            self._display_posn[0] += 1
        else:
            self._display_posn[0] = 0
            self._display_posn[1] += 1
    
    def render_display(self):
        return "\n".join("".join(row) for row in self._display)

    def _start_next_instruction(self):
        """ Takes an instruction, and calls the appropriate implementation method. """
        instruction = self._instructions[self._ip]
        # print(f"Starting instruction: {instruction}")
        
        if instruction[0] == "addx":
            self._doing = [instruction, 2]
        elif instruction[0] == "noop":
            self._doing = [instruction, 1]

        self._ip += 1  # increment the instruction pointer
        if self._ip == len(self._instructions): # we've finished
            self.running_program = False
    
    def _op_addx(self, instruction: tuple):
        """ Takes 2 cycles. Adds val to register x """
        self._x += instruction[-1]

    def _op_noop(self, _: tuple):
        """ Takes 1 cycle. Does nothing. Instruction parameter will be empty. """
    
    def __repr__(self):
        return f"{self.__class__.__name__}(Cycle={self.cycle};x={self._x},pixel={self._display_posn})"




In [116]:
wdir = '/home/alp/Google Drive/Python/AdventCode/2022'
with open (f"{wdir}/10input.txt") as f_in:
    data = f_in.read().splitlines()

interesting_cycles = [20, 60, 100, 140, 180, 220]
signal_strength_sum = 0
crt_computer = CrtComputer(data)
while crt_computer.running_program:
    crt_computer.tick()
    if crt_computer.cycle in interesting_cycles:
        signal_strength_sum += crt_computer.signal_strength

print(signal_strength_sum)
print(crt_computer.render_display())


12540
####.####..##..####.####.#....#..#.####.
#....#....#..#....#.#....#....#..#.#....
###..###..#......#..###..#....####.###..
#....#....#.....#...#....#....#..#.#....
#....#....#..#.#....#....#....#..#.#....
#....####..##..####.####.####.#..#.####.


In [115]:
crt_compter.render_display()

NameError: name 'crt_compter' is not defined

In [103]:
test = [[i for i in range(10)] for j in range(6)]

In [104]:
test

[[0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]

In [105]:
test[0][3]

3